File: Serialize and Deserialize Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        
        if(root == null) return null;

        Queue<TreeNode> queue = new LinkedList();
        queue.add(root);

        StringBuilder result = new StringBuilder();

        while(!queue.isEmpty())
        {
            TreeNode node = queue.poll();

            if(node == null)
            {
                result.append("null");
                result.append(",");
                continue;
            }

            result.append(node.val);
            result.append(",");

            queue.add(node.left);
            queue.add(node.right);
        }

        return result.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        
        if(data == null || data.length() == 0) return null;

        String[] strArray = data.split(",");
        int index = 1;

        Queue<TreeNode> queue = new LinkedList();
        TreeNode root = new TreeNode(Integer.parseInt(strArray[0]));

        queue.add(root);

        while(!queue.isEmpty())
        {
            TreeNode node = queue.poll();

            if(!strArray[index].equals("null"))
            {
                node.left = new TreeNode(Integer.parseInt(strArray[index]));
                queue.add(node.left);
            }

            // We have to even increment when we encounter a null. So we do it outside if
            index++;

            if(!strArray[index].equals("null"))
            {
                node.right = new TreeNode(Integer.parseInt(strArray[index]));
                queue.add(node.right);
            }

            // We have to even increment when we encounter a null. So we do it outside if
            index++;
        }

        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: Serialize_and_Deserialize_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        
        q = deque()        
        q.append(root)
        result = ""
        while q:
            node = q.popleft()
            if node:
                result += str(node.val)
                result += ","
                q.append(node.left)
                q.append(node.right)
                
            else:
                result += "null"
                result += ","

        #print(result)
        return result    
            
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        
        tree = data.split(",")
        #print(tree)        
        q = deque()
        root = TreeNode(tree[0])
        q.append(root)
        i = 1
        while q and i < len(tree):
            node = q.popleft()
            if tree[i] != "null":
                node.left = TreeNode(tree[i])
                q.append(node.left)
            i+=1
            if tree[i] != "null":
                node.right = TreeNode(tree[i])
                q.append(node.right)
            i+=1
          
        return root
        
        
        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
